pkg/utils/pointer/pointer.go

```diff
@@ -17,327 +17,75 @@ limitations under the License.
-	}
-	return 0
-}
-
-// TimeToMetaTime converts a standard Go time.Time to a K8s metav1.Time.
-func TimeToMetaTime(t *time.Time) *metav1.Time {
-	if t == nil {
-		return nil
-	}
-	return &metav1.Time{Time: *t}
-}
-
-// LateInitializeStringPtr returns in if it's non-nil, otherwise returns from
-// which is the backup for the cases in is nil.
-func LateInitializeStringPtr(in *string, from *string) *string {
-	if in != nil {
-		return in
-	}
-	return from
-}
-
-// LateInitializeString returns `from` if `in` is empty and `from` is non-nil,
-// in other cases it returns `in`.
-func LateInitializeString(in string, from *string) string {
-	if in == "" && from != nil {
-		return *from
-	}
-	return in
-}
-
-// LateInitializeTimePtr returns in if it's non-nil, otherwise returns from
-// which is the backup for the cases in is nil.
-func LateInitializeTimePtr(in *metav1.Time, from *time.Time) *metav1.Time {
-	if in != nil {
-		return in
-	}
-	if from != nil {
-		t := metav1.NewTime(*from)
-		return &t
-	}
-	return nil
+	return ptr.Deref(v, 0)
 }
 
 // Int64 converts the supplied int for use with the AWS Go SDK.
-func Int64(v int, o ...FieldOption) *int64 {
-	for _, fo := range o {
-		if fo == FieldRequired && v == 0 {
-			return aws.Int64(int64(v))
-		}
-	}
-
+func ToIntAsInt64(v int) *int64 {
 	if v == 0 {
 		return nil
 	}
-
-	return aws.Int64(int64(v))
+	val64 := int64(v)
+	return &val64
 }
 
-var Int64Ptr = ptr.To[int64]
-
 // Int32 converts the supplied int for use with the AWS Go SDK.
-func Int32(v int, o ...FieldOption) *int32 {
-	for _, fo := range o {
-		if fo == FieldRequired && v == 0 {
-			return aws.Int32(int32(v))
-		}
-	}
-
+func ToIntAsInt32(v int) *int32 {
 	if v == 0 {
 		return nil
 	}
-
-	return aws.Int32(int32(v))
+	val32 := int32(v)
+	return &val32
 }
 
-var Int32Ptr = ptr.To[int32]
-
-// Int64Address returns the given *int in the form of *int64.
-func Int64Address(i *int) *int64 {
-	if i == nil {
-		return nil
-	}
-	return aws.Int64(int64(*i))
-}
-
-var Int64Slice = aws.Int64Slice
-
 // Int32Address returns the given *int in the form of *int32.
-func Int32Address(i *int) *int32 {
-	if i == nil {
+func ToIntAsInt32Ptr(v *int) *int32 {
+	if v == nil {
 		return nil
 	}
-	return aws.Int32(int32(*i))
+	val32 := int32(*v)
+	return &val32
 }
 
-// IntAddress converts the supplied int64 pointer to an int pointer, returning nil if
-// the pointer is nil.
-func IntAddress(i *int64) *int {
-	if i == nil {
+// ToInt32FromIntPtr converts an int32 pointer to an int pointer.
+func ToInt32FromIntPtr(v *int32) *int {
+	if v == nil {
 		return nil
 	}
-	r := int(*i)
-	return &r
+	val := int(*v)
+	return &val
 }
 
-// IntFrom32Address converts the supplied int32 pointer to an int pointer, returning nil if
-// the pointer is nil.
-func IntFrom32Address(i *int32) *int {
-	if i == nil {
+// ToIfNotDefault returns a pointer to val if it does NOT match the default
+// value of T. Otherwise it returns nil.
+func ToIfNotDefault[T comparable](val T) *T {
```

Comment:
I find it a bit hard to remember what exactly this does. Not sure about a better name.

I think "zero value", not default value, is the terminology used in the go documentation, for example: https://tip.golang.org/doc/effective_go


Not sure if `pointer.NilIfZeroValue()` is better :)

Reply:
How about `pointer.ToOrNilIfZeroValue()`?

Reply:
Done